Test Delta Lake metadata handling when schema already exists

pajaks opened the pull request:
## Description
Add test for https://github.com/trinodb/trino/pull/13242 which was merged without a test as we didn't have an infrastructure for such tests at that time.
## Additional context and related issues
https://github.com/trinodb/trino/issues/15282


<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(x) This is not user-visible or docs only and no release notes are required.

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:
```diff
@@ -241,6 +252,37 @@ public void testHideNonDeltaLakeTable()
+    }
+
+    @Test
+    public void testFailWhenCreateTheSameSchemaFromTwoSessions()
```
`testFailWhenCreateTheSameSchemaFromTwoSessions` -> `testCreateSameSchemaFromDifferentSessionsFails`

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:
```diff
@@ -241,6 +252,37 @@ public void testHideNonDeltaLakeTable()
+
+        DeltaLakeMetadata metadata = metadataFactory.create(session.getIdentity());
+
+        metadata.createSchema(session, testSchemaName, Map.of(), TRINO_PRINCIPAL);
+        metadata.createSchema(session, testSchemaName, Map.of(), TRINO_PRINCIPAL);
```
Maybe call `metadata.schemaExists()` after the second call to be clear that the schema exists before doing the second call.

homar replied:
In this and the second test case shouldn't we validate that schema was actually created?

ebyhr replied:
Just calling `DeltaLakeMetadata#createSchema` doesn't trigger the issue in #15177. The situation is remote metastore succeeded, but the client got timed out in **_a_** createSchema operation.

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:
```diff
@@ -241,6 +252,37 @@ public void testHideNonDeltaLakeTable()
     }
 
+    @Test
+    public void testDoubleCreateSchemaForTheSameSession()
```
`testDoubleCreateSchemaForTheSameSession` -> `testCreateSchemaTwiceInTheSameSession`

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/metastore/glue/TestDeltaLakeGlueMetastore.java:
```diff
@@ -241,6 +252,37 @@ public void testHideNonDeltaLakeTable()
+        String testSchemaName = "test_schema_" + randomName();
+
+        DeltaLakeMetadata metadata = metadataFactory.create(session.getIdentity());
+
```
try/finally to ensure that the Glue database does get deleted even in case of exceptions

findinpath commented:
Commit message `Add tests for creating schema` 

Maybe `Test Delta AWS Glue handling when schema already exists`

ebyhr reviewed (commented):
~The new test `testDoubleCreateSchemaForTheSameSession` should fail without commits in https://github.com/trinodb/trino/pull/15177. Currently, it passes. I would recommend reverting commits temporarily when writing the test.~ I realized the issue number in the description was wrong.

pajaks commented:
> 

It was mistake in description. Tests are for [#13242](https://github.com/trinodb/trino/pull/13242)

ebyhr commented:
> Tests are for https://github.com/trinodb/trino/pull/13242

The PR isn't specific to Glue. Could you explain why this PR add tests only for Glue?

pajaks commented:
> > Tests are for #13242
> 
> The PR isn't specific to Glue. Could you explain why this PR add tests only for Glue?

Test was in wrong place. I moved it to generic tests and changed title.

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeMetadata.java:
```diff
@@ -465,6 +473,35 @@ public void testGetInputInfoForUnPartitionedTable()
     }
 
+    @Test
+    public void testCreateSchemaTwiceInTheSameSession()
```
This feels like a unit test and somebody without a context for the `SchemaAlreadyExistsException` problem would have a hard time to figure out what are the tests there.
What about using a specific test - similar to `TestIcebergFileMetastoreTableOperationsInsertFailure` ?

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:
```diff
@@ -0,0 +1,110 @@
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.lang.String.format;
+
+public class TestDeltaLakeSchemaCreate
```
How about `TestDeltaLakeCreateSchemaInternalRetry` or something?

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:
```diff
@@ -0,0 +1,110 @@
+            public synchronized void createDatabase(Database database)
+            {
+                super.createDatabase(database);
+                afterCreateDatabaseFunction.run();
```
I'm not sure why `Runnable` is needed. I would throw an exception directly here.

pajaks replied:
My idea was to make test itself more self describing, but I agree that it my be unnecessary complication.

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:
```diff
@@ -0,0 +1,110 @@
+    @Test
+    public void testSchemaCreationWithTimeout()
+    {
+        // Simulate retry mechanism with timeout failure.
+        // 1. createDatabase correctly create schema but timeout is triggered
+        // 2. Retry to createDatabase throws SchemaAlreadyExistsException
```
Move this comment to before `super.createDatabase(database);` if you inline `Runnable`.

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:
```diff
@@ -0,0 +1,110 @@
+        // 1. createDatabase correctly create schema but timeout is triggered
+        // 2. Retry to createDatabase throws SchemaAlreadyExistsException
+        afterCreateDatabaseFunction = () -> {
+            throw new SchemaAlreadyExistsException("Test-simulated metastore schema already exist exception");
```
The argument takes a schema name, not an error message.

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:
```diff
@@ -0,0 +1,110 @@
+            throw new SchemaAlreadyExistsException("Test-simulated metastore schema already exist exception");
+        };
+
+        assertQuerySucceeds(format("CREATE SCHEMA %s", TEST_SCHEMA));
+        assertQuery(format("SHOW SCHEMAS LIKE '%s'", TEST_SCHEMA), format("VALUES ('%s')", TEST_SCHEMA));
```
String concatenation provides SQL highlight. We use this style when the query is simple.
```java
        assertQuerySucceeds("CREATE SCHEMA " + TEST_SCHEMA);
        assertQuery("SHOW SCHEMAS LIKE '" + TEST_SCHEMA + "'", "VALUES '" + TEST_SCHEMA + "'");
```

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreate.java:
```diff
@@ -0,0 +1,110 @@
+
+        assertQuerySucceeds(format("CREATE SCHEMA %s", TEST_SCHEMA));
+        assertQuery(format("SHOW SCHEMAS LIKE '%s'", TEST_SCHEMA), format("VALUES ('%s')", TEST_SCHEMA));
+    }
```
Optional: Ideally, we should add another test method to verify that `CREATE SCHEMA` throws an exception when `presto_query_id` is different from the session query id.

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeMetadata.java:
```diff
@@ -465,6 +473,35 @@ public void testGetInputInfoForUnPartitionedTable()
+                .isInstanceOf(SchemaAlreadyExistsException.class)
+                .hasMessage(format("Schema already exists: '%s'", testSchemaName));
+        assertThat(deltaLakeMetadata.schemaExists(SESSION, testSchemaName)).isTrue();
+    }
```
Can we revert this change in this class or add some comments for clarifying the test purpose?

ebyhr reviewed (commented):
Almost good to me. Thanks for adding this test.

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSchemaCreateInternalRetry.java:
```diff
@@ -0,0 +1,129 @@
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+
+public class TestDeltaLakeSchemaCreateInternalRetry
```
```suggestion
public class TestDeltaLakeCreateSchemaInternalRetry
```

ebyhr commented:
> Add tests for creating schema

@pajaks Could you fix the commit title as "Add tests for internal retry to create schema in Delta Lake"?

ebyhr commented:
@cla-bot check

ebyhr commented:
@pajaks Are you going to add tests for Hive as well?

pajaks commented:
> @pajaks Are you going to add tests for Hive as well?

It seems like a good idea. I will look into it.
